README.md
## Description
Goal: This project is a start to make an application with which we can map or highlight safe and or unsafe areas in an area. 

Present: An API with which live data changes a threeJS model to represent safe or unsafe places and the speed with which the user moves at that time. 
Features: Upon updating the data, the threejs model changes speed and color according to the latest data.

## Getting Started
The back-end is a dockerised node.js application made with express.

The front-end is made with ThreeJS.

Have a look at our 'contribution guidelines' if you want to contribute to the project.

# How to set up a local development environment:

1. Clone the repository:

  ```bash
  git clone https://github.com/RebeccaRosies/DevelopmentV.git
  ```
2. Change the .env.template to .env and fill in with your respective data
3. Change the database URL in dbConnection.js to your own database

4. Run docker-compose, this takes care of installing dependencies, building packages and ensuring your workspace is dev-ready.

  ```bash
 docker-compose build
 docker-compose up
  ```
  
5. To push your own version of the image to docker hub run: 

  ```bash
  docker-compose -f docker-compose.prod.yml up --build 
    ```
    
6. To run the testing with jest and super test: 

  ```bash
 cd images/api
 npm init -y
 npm run test
    ```

7. Go to the ThreeJS folder, open a new terminal and install and run node:

  ```bash
npm init -y
npm run

  ```

That's it, you are good to go! 

## Error Handling
If you run into errors you may contact me at rebecca.rosies@student.ehb.be.
Alternatively, feel free to **file a new issue** with a respective title and description on the the [RebeccaRosies/DevelopmentV](https://github.com/RebeccaRosies/DevelopmentV/issues) repository

## Status
The project is still in development.




CHANGELOG.md
# ThreeJS safeZone app
## 0.1
* Initial release.


images/api/src/__tests__/apiGet.test.js
const supertest = require("supertest")
const s = require("../server.js")


//Tests the GET ENDPOINT

describe('GET /dataLog', () => {
  beforeAll(async () => {
    //connect to database
    app = await s();
  });
  
  it('tests if the get endpoint makes a succesful connection', async() => {
     const request = supertest(app);

      const res = await request
        .get('/dataLog')
        .set('Accept', 'application/json')

        expect(typeof res.body).toBe('object');
        expect(res.statusCode).toBe(200);
    });

    it('tests if the appropiate error is sent when the endpoint is incorrect', async() => {
       const request = supertest(app);
  
        const res = await request
          .get('/data')
          .set('Accept', 'application/json')
  
          expect(typeof res.body).toBe('object');
          expect(res.statusCode).toBe(404);
      });

  })



images/api/src/__tests__/apiPost.test.js
const supertest = require("supertest")
const s = require("../server.js")


//Tests the POST ENDPOINT

describe('POST /dataLog', () => {
    beforeAll(async () => {
        //connect to database
        app = await s();
      });

    it('tests if the post endpoint posts an object', async () => {

        const request = supertest(app);
        const res = await request
            .post('/dataLog')
            .set('Accept', 'application/json')
            .send({
                speed: 2.1,
                safe: true
            })

        expect(typeof res.body).toBe('object');
        expect(res.statusCode).toBe(200);
    });

    it('tests if the appropriate error is send when the endpoint is incorrect', async () => {

        const request = supertest(app);
        const res = await request
            .post('/data')
            .set('Accept', 'application/json')
            .send({
                speed: 8.4,
                safe: true
            })

        expect(typeof res.body).toBe('object');
        expect(res.statusCode).toBe(404);
    });

    it('tests if the appropriate error is send if the datatype for speed is wrong', async () => {

        const request = supertest(app);
        const res = await request
            .post('/dataLog')
            .set('Accept', 'application/json')
            .send({
                speed: "twee",
                safe: true
            })

        expect(typeof res.body).toBe('object');
        expect(res.statusCode).toBe(400);
        expect(res.text).toEqual('ERROR: speed should be a number')
    });

    it('tests if the appropriate error is send if the datatype for safe is wrong', async () => {

        const request = supertest(app);
        const res = await request
            .post('/dataLog')
            .set('Accept', 'application/json')
            .send({
                speed: 5.3,
                safe: "true"
            })

        expect(typeof res.body).toBe('object');
        expect(res.statusCode).toBe(400);
        expect(res.text).toEqual('ERROR: safe should be a boolean')
    });

    it('tests if the appropriate error is send if the dataobject is empty', async () => {

        const request = supertest(app);
        const res = await request
            .post('/dataLog')
            .set('Accept', 'application/json')
            .send({
            })

        expect(typeof res.body).toBe('object');
        expect(res.statusCode).toBe(400);
        expect(res.text).toEqual('ERROR: missing speed or safe')
    });


})

images/ThreeJS_WebContent/main.js
const scene = new THREE.Scene();
const camera = new THREE.PerspectiveCamera(75, window.innerWidth / window.innerHeight, 0.1, 1000);

const renderer = new THREE.WebGLRenderer();
renderer.setSize(window.innerWidth, window.innerHeight);
document.body.appendChild(renderer.domElement);

const geometry = new THREE.BoxGeometry(1, 1, 1);
const material = new THREE.MeshBasicMaterial({
    color: 0x00ff00
});
const cube = new THREE.Mesh(geometry, material);
scene.add(cube);

camera.position.z = 5;

/* --------------------------------------------------- */
function getData(){
    console.log("data");
    fetch('mongodb+srv://Admin:rebeccarosiesadmin@cluster0.dzxmp.mongodb.net/dataLog', {
        method: 'GET',
        headers: {
            'Content-Type': 'application/json',
        }
    })
    .then(response => response.json())
    .then(data => {
        console.log(data);
    }); 
}

getData();


function animate() {
    requestAnimationFrame(animate);
    renderer.render(scene, camera);
}
animate();

images/ThreeJS_WebContent/WebGL.js
class WebGL {

	static isWebGLAvailable() {

		try {

			const canvas = document.createElement('canvas');
			return !!(window.WebGLRenderingContext && (canvas.getContext('webgl') || canvas.getContext('experimental-webgl')));

		} catch (e) {

			return false;

		}

	}

	static isWebGL2Available() {

		try {

			const canvas = document.createElement('canvas');
			return !!(window.WebGL2RenderingContext && canvas.getContext('webgl2'));

		} catch (e) {

			return false;

		}

	}

	static getWebGLErrorMessage() {

		return this.getErrorMessage(1);

	}

	static getWebGL2ErrorMessage() {

		return this.getErrorMessage(2);

	}

	static getErrorMessage(version) {

		const names = {
			1: 'WebGL',
			2: 'WebGL 2'
		};

		const contexts = {
			1: window.WebGLRenderingContext,
			2: window.WebGL2RenderingContext
		};

		let message = 'Your $0 does not seem to support <a href="http://khronos.org/webgl/wiki/Getting_a_WebGL_Implementation" style="color:#000">$1</a>';

		const element = document.createElement('div');
		element.id = 'webglmessage';
		element.style.fontFamily = 'monospace';
		element.style.fontSize = '13px';
		element.style.fontWeight = 'normal';
		element.style.textAlign = 'center';
		element.style.background = '#fff';
		element.style.color = '#000';
		element.style.padding = '1.5em';
		element.style.width = '400px';
		element.style.margin = '5em auto 0';

		if (contexts[version]) {

			message = message.replace('$0', 'graphics card');

		} else {

			message = message.replace('$0', 'browser');

		}

		message = message.replace('$1', names[version]);

		element.innerHTML = message;

		return element;

	}

}

// module.exports = WebGL;

images/ThreeJS_WebContent/src/index.js


const scene = new THREE.Scene();
const camera = new THREE.PerspectiveCamera(75, window.innerWidth / window.innerHeight, 0.1, 1000);

const renderer = new THREE.WebGLRenderer();
renderer.setSize(window.innerWidth, window.innerHeight);
document.body.appendChild(renderer.domElement);

const geometry = new THREE.BoxGeometry(1, 1, 1, 1, 1, 1);

/* 
//Shader Material //
//The material code is based on the code from https://www.youtube.com/watch?v=EntBBM6nqQA //
*/
const cubeMaterialSafe = new THREE.ShaderMaterial({
    wireframe:true,
    vertexShader: `
    void main()	{
      // projectionMatrix, modelViewMatrix, position -> passed in from Three.js

      vec4 result;
      result = vec4(position.x, sin(position.z/3.0)+position.y, position.z, 1.0);

      gl_Position = projectionMatrix
        * modelViewMatrix
        * result;
    }
    `,
    fragmentShader: `
    void main() {
      gl_FragColor = vec4(0.0, 1.0, 0.0, 1.0);
    }
    `,
});

const cubeMaterialUnsafe = new THREE.ShaderMaterial({
    wireframe:true,
    vertexShader: `
    void main()	{
      // projectionMatrix, modelViewMatrix, position -> passed in from Three.js

      vec4 result;
      result = vec4(position.x, sin(position.z/4.0)+position.y, position.z, 1.0);

      gl_Position = projectionMatrix
        * modelViewMatrix
        * result;
    }
    `,
    fragmentShader: `
    void main() {
      gl_FragColor = vec4(1.0, 0.0, 0.0, 1.0);
    }
    `,
});

const cubeMaterialDefault = new THREE.ShaderMaterial({
    wireframe:true,
    vertexShader: `
    void main()	{
      // projectionMatrix, modelViewMatrix, position -> passed in from Three.js

      vec4 result;
      result = vec4(position.x, sin(position.z/5.0)+position.y, position.z, 1.0);

      gl_Position = projectionMatrix
        * modelViewMatrix
        * result;
    }
    `,
    fragmentShader: `
    void main() {
      gl_FragColor = vec4(0.0, 0.0, 1.0, 1.0);
    }
    `,
});


let latestSafe = true;

//Basic materials (no shaders)
/* let materialSafe = new THREE.MeshBasicMaterial({
    color: 0x00ff00,
    wireframe:true
}); 
let materialUnsafe = new THREE.MeshBasicMaterial({
    color: 0xff0000,
    wireframe:true
}); 
let materialDefault = new THREE.MeshBasicMaterial({
    color: 0x0000ff,
    wireframe:true
});  */

let cube = new THREE.Mesh(geometry, cubeMaterialSafe);


scene.add(cube);

const light = new THREE.AmbientLight(0x404040); // soft white light
scene.add(light);

camera.position.z = 5;

/* --------------------------------------------------- */

/** 
* @returns an Array with objects containing: 
*           speed = (Number) = Ex: 2.1 = How fast was a person going at a certain time and place 
*           safe = (Boolean) = EX: true = Did the person feel safe
 */
function getData() {
    console.log("data");
    fetch('http://localhost/dataLog', {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json',
            }
        })
        .then(response => response.json())
        .then(data => {
            getSpeed(data);
        });
};
getData();

let speedArray = [];
let timesArray = [];
let safeArray = [];
let latestspeed = 0;


function getSpeed(data) { 
    data.forEach((date, i) => {
        speedArray.push(date.speed);
        timesArray.push(i * 5);
        safeArray.push(date.safe)
    });
    console.log(speedArray);
    console.log(speedArray[(speedArray.length-1)]);
    console.log(timesArray);
    console.log(safeArray);
 
    latestspeed = speedArray[(speedArray.length-1)];
    latestSafe = safeArray[(safeArray.length-1)];
    
}
/* 
const track1  = new NumberKeyframeTrack("track1", timesArray, speedArray, THREE.InterpolateLinear);
const tracks = [track1];
const clip = new AnimationClip("speed", -1, tracks)
const mixer = new THREE.AnimationMixer(cube);
const clips = cube.animations;

function update(){
    mixer.update(1);
}

const animation = new AnimationClip("cubespeed", -1, [track1] ); */


function animate(time) {
     // update the speed of the cube whenever the latest data comes through
    let goSpeed = speedArray[(speedArray.length-1)];
    console.log(latestSafe);
    requestAnimationFrame(animate);
    cube.rotation.x = goSpeed * time / 1000;
    cube.rotation.y = 0; 

    // update the color of the cube whenever the latest data comes through
    if(latestSafe == NaN){
        cube.material = cubeMaterialDefault; 
    } else if (latestSafe == true) {
         cube.material = cubeMaterialUnsafe;
    } else if (latestSafe == false) {
         cube.material = cubeMaterialUnsafe; 
    }
    
    //render the scene
    renderer.render(scene, camera);
}
animate();

images/api/src/server.js
require('dotenv').config({
    path: './.env'
});

const {
    ObjectId
} = require('mongodb');

const {
    connectToDb,
    client
} = require('./dbConnection');

const collections = {}

const setup = () =>
    new Promise((resolve, reject) => { 

        connectToDb()
            .then((collection) => {

                collections["les1"] = collection;
                resolve(app)
            })
            .catch((error)=>{
                console.error(error)
            })
     }) 
 

const express = require('express');
const bodyParser = require("body-parser");

const app = express();

//app.use(cors({origin: '*'}));
app.use(function(req, res, next) {
    res.header("Access-Control-Allow-Origin", "*");
    res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
    next();
  });


app.use(bodyParser.urlencoded({
    extended: true
}))
app.use(bodyParser.json())

/** 
* @returns hi world = (String)
 */
app.get("/", (request, response) => {
    response.send({
        message: "hi world"
    })
})

/** 
* @returns speed = (Number) = Ex: 2.1 = How fast was a person going at a certain time and place 
* @returns safe = (Boolean) = EX: true = Did the person feel safe
 */
app.get("/dataLog", async (req, res) => {
    try {
        const myData = await collections["les1"].find({}).toArray(); // Find document & convert it to an array
        console.log(myData); // Print to the console
        //res.status(200).send(myData); //Send back the data with the response
        
        try{
            res.status(200).send(myData)
        } catch (error){
            res.status(400).send({message:'your data was not found'})
        }
    
    } catch (err) {
        console.log('error', err);
        res.status(500).send({
            error: 'an error has occured'
        });
    } 
})

/** 
* @param speed = (Number) = Ex: 2.1 = How fast was a person going at a certain time and place 
* @param safe = (Boolean) = EX: true = Did the person feel safe 
 */
app.post('/dataLog', async (req, res) => {
    try {
        if (req.body.speed == undefined || req.body.safe == undefined) {
            res.status(400).send('ERROR: missing speed or safe');
            return;
        }
        else if (typeof req.body.speed != 'number'){
            console.log("ERROR STRING");
            res.status(400).send('ERROR: speed should be a number');
            return;
        }
        else if (typeof req.body.safe != 'boolean'){
            console.log("ERROR STRING");
            res.status(400).send('ERROR: safe should be a boolean');
            return;
        }
       console.log(req.body.speed);

        //construct a document
        let newData = {
            speed: req.body.speed,
            safe: req.body.safe
        }

        console.log(newData)
        //insert into database
        let insertResult = await collections["les1"].insertOne(newData);
        //console.log(insertResult)
        res.send(insertResult)
        return;

    } catch (error) {
        console.log("error", error);
        res.status(500).send({
            error: 'an error has occured',
            value: error
        });
    }
})


/* deletes all data */
app.delete('/dataLog', async (req, res) => {
    try {
        const message = {
            deleted: "All data deleted"
        }
        const result = await collections["les1"].deleteMany();
        if (result.deletedCount >= 1) {
            res
                .status(200)
                .send(message);
        } else {
            res
                .status(404)
                .send("No documents matched the query. Deleted 0 documents.");
        }
    } catch (err) {
        console.log('error');
        res.status(500).send({
            error: 'an error has occured',
        });
    }
})

/** 
* @param speed = (Number) = Ex: 2.1 = How fast was a person going at a certain time and place 
* @param safe = (Boolean) = EX: true = Did the person feel safe 
* @returns speed (Number) = Ex: 2.1 = How fast was a person going at a certain time and place
* @returns safe (Boolean) = EX: true = Did the person feel safe 
 */
app.put("/dataLog/:id", async (req, res) => {
    // check for body data
    const error = {
        error: "Bad request",
        value: "Missing speed or safe"
    }

    try {
        if (req.body.speed == undefined || req.body.safe == undefined) {
            res.status(400).send(error);
            return;
        }

        // Create a query for data to update
        const query = {
            _id: ObjectId(req.params.id)
        };
        const message = {
            deleted: "Data updated"
        }

        // update the data
        const updateData = {
            speed: req.body.speed,
            safe: req.body.safe
        };
        console.log(query, updateData);
        // Updating the data
        const result = await collections["les1"].updateOne(query, {
            $set: updateData
        });

        // Send back success message
        res.status(201).send(result);
    } catch (error) {
        console.log(error);
        res.status(500).send({
            error: "something went wrong",
        });
    }
});


module.exports = setup;
